### PR TITLE
respect keepkernel on exit for zmqshell in console

### LINKED ISSUE #8482: quit(keep_kernel=True) doesn't work
I'm using IPython 3.1.0

### PR DESCRIPTION
**Pull request by Carreau**
closes #8482

---

this is against 3x.

### CODE REVIEW

**Comment by takluyver** [comment]
This looks like it will stop the frontend from exiting if `keep_kernel` is True. That's not what should happen - the frontend should exit and leave the kernel running.

**Comment by Carreau** [comment]
Ok, what about that ?

**Comment by takluyver** [comment]
I think this now looks functionally correct, but:

- I wouldn't change anything about the Ctrl-C handling. Ctrl-D is the shortcut to quit.
- I'd like the logic to be simpler overall before we merge it.

**Comment by Carreau** [comment]
> I wouldn't change anything about the Ctrl-C handling. Ctrl-D is the shortcut to quit.

Ctrd-D is not a signal, and does not raise, so its more complicated to handle, Hand here it is really to kill the program when it's stuck. Ctrl-D already exit if you are connected to a kernel. But here it was to escape the infinite loop if you `--existing` and there is no kernels. Still removed though.

I cannot simplify the logic much more, I removed the `ask_exit`. I can make own_kernel not a property if you like.

**Comment by takluyver** [comment]
Ctrl-D at an input prompt should raise EOFError.

File: IPython/terminal/console/interactiveshell.py [hunk 1]
```diff
@@ -115,6 +115,14 @@ def _client_changed(self, name, old, new):
         self.session_id = new.session.session
     session_id = Unicode()
 
+    @property
+    def own_kernel(self):
+        return self._own_kernel
+
+    @own_kernel.setter
+    def own_kernel(self, b):
+        self._own_kernel = bool(b)
```
Comment (takluyver) [inline IPython/terminal/console/interactiveshell.py]: This property does seem redundant. The only two places you're setting it, it's a boolean already.

**Comment by Carreau** [comment]
But you are stuck before.I suppose when you `console --existing` you are actually stuck waiting for the kernel to respond, so EOF does nothing, anyway, that's a separate issue.

File: IPython/terminal/console/interactiveshell.py (hunk 1 above)
Comment (takluyver) [inline IPython/terminal/console/interactiveshell.py]: s/two places/one place/

**Comment by takluyver** [comment]
With the property gone, I think this looks OK now.

**Comment by Carreau** [comment]
property gone

**Comment by zer0n** [comment]
`Ctrl-D` throws 

```
In [7]:
Do you really want to exit ([y]/n)? y
Traceback (most recent call last):
...
  File ".../site-packages/IPython/terminal/console/interactiveshell.py", line 426, in mainloop
    if self.keepkernel and not self.own_kernel:
AttributeError: 'ZMQTerminalInteractiveShell' object has no attribute 'keepkernel'
```

This is minor, as it doesn't kill the kernel. However, graceful exit would be nicer than throwing exceptions.

**Comment by takluyver** [comment]
@Carreau, can you look into that?

**Comment by Carreau** [comment]
Yep will do .

**Comment by Carreau** [comment]
`IPython/terminal/console/interactiveshell.py` does not exist anymore. Are you on IPython master ? and `jupyter_console` master ?

**Comment by zer0n** [comment]
Nope. Actually, I was using IPython 3.1.0. I updated the changed files manually.

**Comment by Carreau** [comment]
Ah, yes this has been backported. sorry, let me retry on 3.x

**Pull request merged**